From 8c6d1cfca22f87b20b79a419ceabdb182846ff4a Mon Sep 17 00:00:00 2001
From: progier389 <progier@redhat.com>
Date: Tue, 27 Jan 2026 15:39:39 +0100
Subject: [PATCH] Issue 7206 - Should log whether TLS key is PQC or not (#7207)

Append [PQC] to the cipher name when logging SSL/TLS if the Key Exchange is one of the KEM group
If connection debug level is enabled, logs in error log the keaType and keaGroup (as integer)

Issue: #7206

Reviewed by: @tbordaz, @droideck (Thanks!)

* Issue 7206 - Should log whether TLS key is PQC or not

* Fix Sourcery A/I comments

* Check PQC in test case

* Update ldap/servers/slapd/auth.c

Co-authored-by: Simon Pichugin <spichugi@redhat.com>

---------

Co-authored-by: Simon Pichugin <spichugi@redhat.com>
---
 dirsrvtests/tests/suites/tls/mldsa_test.py |  7 ++++
 ldap/servers/slapd/auth.c                  | 45 +++++++++++++++++++++-
 ldap/servers/slapd/ssl.c                   |  2 +-
 3 files changed, 52 insertions(+), 2 deletions(-)

diff --git a/dirsrvtests/tests/suites/tls/mldsa_test.py b/dirsrvtests/tests/suites/tls/mldsa_test.py
index 2c815088b..d6cbaf662 100644
--- a/dirsrvtests/tests/suites/tls/mldsa_test.py
+++ b/dirsrvtests/tests/suites/tls/mldsa_test.py
@@ -274,6 +274,7 @@ def test_mldsa(topo):
     """
 
     inst = topo.standalone
+    inst.config.set('nsslapd-errorlog-level', str(16384 + 8))
     inst.enable_tls()
 
     cm = CertmapLegacy(inst)
@@ -293,6 +294,9 @@ def test_mldsa(topo):
             'loginShell': '/bin/false',
             'description': cert_dn })
 
+    inst.config.set("nsslapd-accesslog-logbuffering", "off")
+    inst.config.set("nsslapd-errorlog-logbuffering", "off")
+
     tmpdir_kwargs = {}
     if sys.version_info >= (3, 12):
         tmpdir_kwargs['delete'] = not DEBUGGING
@@ -313,6 +317,9 @@ def test_mldsa(topo):
         res.check_returncode()
         # If ldapsearch is successful then defaultnamingcontext should be in res.stdout
         assert "defaultnamingcontext" in res.stdout
+    assert inst.ds_access_log.match('.*RESULT.*dn="uid=test_user,ou=people,dc=example,dc=com".*')
+    assert inst.ds_access_log.match('.*TLS.*[PQC].*')
+    assert inst.ds_error_log.match('.*check_pqc.*')
 
 
 if __name__ == '__main__':
diff --git a/ldap/servers/slapd/auth.c b/ldap/servers/slapd/auth.c
index 48e4b7129..9e83408d4 100644
--- a/ldap/servers/slapd/auth.c
+++ b/ldap/servers/slapd/auth.c
@@ -395,6 +395,43 @@ handle_bad_certificate(void *clientData, PRFileDesc *prfd)
 }
 
 
+/*
+ * Determine if the connection key exchange is Post Quantum Cryptography aware.
+ * This function may need to evolve with NSS as more PQC methods get supported.
+ */
+static bool
+check_pqc(uint64_t connid, SSLChannelInfo *sci)
+{
+    /*
+     * FYI: To interpret the values:
+     * KeaType and KeaGroup values are defined in
+     * https://github.com/nss-dev/nss/blob/master/lib/ssl/sslt.h
+     * Respectively in SSLKEAType and SSLNamedGroup enums
+     */
+    slapi_log_err(SLAPI_LOG_CONNS, "check_pqc", "conn=%" PRIu64 " TLS keaType=%d keaGroup=%d\n",
+                  connid, sci->keaType, sci->keaGroup);
+#ifdef MAX_ML_DSA_PRIVATE_KEY_LEN
+    /* NSS supports PQC (because of the ifdef). Now lets check if the connection uses it */
+    /* Check that PQC KeaType is hybrid */
+    switch (sci->keaType) {
+        case ssl_kea_ecdh_hybrid:
+        case ssl_kea_ecdh_hybrid_psk:
+            break;
+        default:
+            return false;
+    }
+    /* Check that PQC keaGroup is KEM */
+    switch (sci->keaGroup) {
+        case ssl_grp_kem_secp256r1mlkem768:
+        case ssl_grp_kem_secp384r1mlkem1024:
+        case ssl_grp_kem_mlkem768x25519:
+        case ssl_grp_kem_xyber768d00:
+            return true;
+    }
+#endif
+    return false;
+}
+
 /*
  * Get an identity from the client's certificate (if any was sent).
  *
@@ -417,6 +454,7 @@ handle_handshake_done(PRFileDesc *prfd, void *clientData)
     SSLCipherSuiteInfo cipherInfo;
     char *subject = NULL;
     char sslversion[64];
+    bool pqc = false;
     int err = 0;
 
     if ((slapd_ssl_getChannelInfo(prfd, &channelInfo, sizeof(channelInfo))) != SECSuccess) {
@@ -454,7 +492,12 @@ handle_handshake_done(PRFileDesc *prfd, void *clientData)
     }
 
     keySize = cipherInfo.effectiveKeyBits;
-    cipher = slapi_ch_strdup(cipherInfo.symCipherName);
+    pqc = check_pqc(conn->c_connid, &channelInfo);
+    if (pqc) {
+        cipher = slapi_ch_smprintf("%s[PQC]", cipherInfo.symCipherName);
+    } else {
+        cipher = slapi_ch_strdup(cipherInfo.symCipherName);
+    }
 
     /* If inside an Start TLS operation, perform the privacy level discovery
      * and if the security degree achieved after the handshake is not reckoned
diff --git a/ldap/servers/slapd/ssl.c b/ldap/servers/slapd/ssl.c
index 053db5424..7d5db2cdd 100644
--- a/ldap/servers/slapd/ssl.c
+++ b/ldap/servers/slapd/ssl.c
@@ -748,7 +748,7 @@ SSLPLCY_Install(void)
     if (!slapd_pk11_isFIPS()) {
         /* Set explicitly PQC algorithm policy if it is not set by default */
         for (size_t i=0; s == SECSuccess && i < PR_ARRAY_SIZE(oids); i++) {
-            int oflags = 0;
+            PRUint32 oflags = 0;
             (void) NSS_GetAlgorithmPolicy(oids[i], &oflags);
             if ((oflags & flags) != flags) {
                 s = NSS_SetAlgorithmPolicy(oids[i], flags, 0);
-- 
2.52.0

